### PR TITLE
db: adds mid txn errors

### PR DESCRIPTION
**Pull request by sanderpick**
Right now, when you call methods on a transaction like `Save`, `Delete`, etc., an error will end the txn, which is cumbersome in cases where you want to handle that error and retry (w/o recreating the whole transaction).

Easiest solution: Add an error to all the reply objects (CreateReply, etc.)

### CODE REVIEW

File: net/net.go [hunk 1]
```diff
@@ -386,7 +386,6 @@ func (n *net) PullThread(ctx context.Context, id thread.ID, opts ...core.ThreadO
 }
 
 func (n *net) pullThread(ctx context.Context, id thread.ID) error {
-	log.Debugf("pulling thread %s...", id)
```
Comment (sanderpick) [inline net/net.go:389]: This creates too much output, to the point where you can't easily make sense of the logs.

File: db/collection.go [hunk 2]
```diff
@@ -619,6 +619,17 @@ func (t *Txn) Discard() {
 	t.discarded = true
 }
 
+// RefreshCollection updates the transaction's collection reference from the master db map,
+// which may have received updates while the transaction is open.
+func (t *Txn) RefreshCollection() error {
```
Comment (sanderpick) [inline db/collection.go:624]: This is needed to pick up collection changes during a transaction.

**Comment by sanderpick** [comment]
Oh, note that the new errors are only used during a transaction... normal non-txn calls return the error as before.

File: db/collection.go (hunk 2 above)
Comment (carsonfarmer) [reply db/collection.go:624]: This seems odd given we're using a transaction here :P But I guess this actually makes sense, given our transactions provide atomicity but not isolation guarantees!

**Review by carsonfarmer — APPROVED** [review]
This is great @sanderpick, exactly what we need for remote offline-frist updates tbh.

File: db/collection.go (hunk 2 above)
Comment (sanderpick) [reply db/collection.go:624]: Yeah, it's kinda weird :/ Though, the transaction changes themselves and the underlying state are still "locked" during the transaction, as you stated differently above. Really I just needed the ability to update the collection after catching a schema mismatch error _during the txn_.

**Comment by sanderpick** [comment]
Thanks for the late night look!

**Comment by bitcard** [comment]
mark

**Pull request merged**